src/components/structures/RoomDirectory.tsx

```diff
@@ -122,27 +122,33 @@ export default class RoomDirectory extends React.Component<IProps, IState> {
                 const myHomeserver = MatrixClientPeg.getHomeserverName();
                 const lsRoomServer = localStorage.getItem(LAST_SERVER_KEY);
                 const lsInstanceId = localStorage.getItem(LAST_INSTANCE_KEY);
-                const configSevers = SdkConfig.get().roomDirectory?.servers || [];
-                const settingsServers = SettingsStore.getValue("room_directory_servers") || [];
-                const roomServer = [...configSevers, ...settingsServers].includes(lsRoomServer)
-                    ? lsRoomServer
-                    : myHomeserver;
-                const instanceIds = [];
-                if (roomServer === myHomeserver) {
-                    Object.values(this.protocols).forEach((protocol) => {
-                        protocol.instances.forEach((instance) => instanceIds.push(instance.instance_id));
-                    });
+                const roomServer = (
+                    SdkConfig.get().roomDirectory?.servers?.includes(lsRoomServer) ||
+                    SettingsStore.getValue("room_directory_servers")?.includes(lsRoomServer)
+                ) ? lsRoomServer : myHomeserver;
+
+                let instanceId: string = null;
+                if (
+                    (lsInstanceId === ALL_ROOMS) ||
+                    (
+                        roomServer === myHomeserver &&
+                        Object.values(this.protocols).some(p => p.instances.some(i => i.instance_id === lsInstanceId))
+                    )
+                ) {
```

Comment:
actually, I misled you!

`roomServer === myHomeserver` is needed for both cases

```suggestion
                if (roomServer === myHomeserver && (
                	lsInstanceId === ALL_ROOMS) ||
                    Object.values(this.protocols).some(p => p.instances.some(i => i.instance_id === lsInstanceId))
                )) {
```

Reply:
/me should actually read the code before copy-pasting it...